src/Symfony/Component/Validator/Constraints/ExpressionLanguageSyntaxValidator.php

```diff
@@ -39,6 +39,10 @@ public function validate($expression, Constraint $constraint): void
             throw new UnexpectedTypeException($constraint, ExpressionLanguageSyntax::class);
         }
 
+        if (true === $constraint->allowNullAndEmptyString && (null === $expression || '' === $expression)) {
+            return;
```

Comment:
The deprecation should be triggered here instead of the constraint constructor to ensure runtime context, otherwise it may be lost in the app cache warm up, something like:
```php
if (null === $expression || '' === $expression) {
    if ($constraint->allowNullAndEmptyString) {
        return;
    }

    trigger_deprecation(...);
}

Reply:
I'll wait for other comments about it, because all triggered deprecations in Constraints are in the Constraint class and not in the Validator (eg: Length, Range)